### PR TITLE
Parse negative integer literals correctly

### LINKED ISSUE #618: lhist negative ranges
lhist accepts negative numbers but drops the sign internally, which leads to unexpected behaviour:

```
root@ubuntu-bionic:~/build# sudo ../bpftrace -e 'i:ms:1 { @=lhist(-10, -10, 100, 1); }'
lhist() too many buckets, must be <= 1000 (would need 464824020)
lhist() min must be less than max (provided min 93943354841488 and max 100)
lhist() step is too large for the given range (provided step 1 for range -93943354841388)
```
```
root@ubuntu-bionic:~/build# sudo ../bpftrace -e 'i:ms:1 { @=lhist(-10, -10, -1, 1); }'
Attaching 1 probe...
^C

@:
[103233936, 103233937)      91 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
```

```
sudo ../bpftrace -e 'i:ms:1 { @=lhist(-10, 0, 10, -1); }'
lhist() step is too large for the given range (provided step 94564705045168 for range 10)
```

What I'd expect would be something like:

```
sudo ../bpftrace -e 'i:ms:1 { @=lhist(-10, -10, 10, 1); }'
lhist(): range must be postive (min >0; max > min)`
```

### PR DESCRIPTION
**Pull request by danobi**
The parser was parsing "-10" an a unary op on an integer. This
led to reading invalid memory because the semantic analyser
did a static_cast upcast to `Integer` from `Expression`
(which was actually an `Unop`).

This patch requires the constant lhist arguments to be positive
integer literals. This patch also modifies the parser so it correctly
parses negative inteters as literals (vs `Unop`s).

Test Plan:
Before:
```
$ sudo ./build/src/bpftrace -e 'i:ms:1 { @=lhist(-10, -10, 10, 1); }'
lhist() min must be less than max (provided min 13989904 and max 10)
lhist() step is too large for the given range (provided step 1 for range
-13989894)
```

After:
```
$ sudo ./build/src/bpftrace -e 'i:ms:1 { @=lhist(-10, -10, 10, 1); }'
lhist() min must be non-negative (provided min -10)
```

This closes #618.

### CODE REVIEW

**Comment by ajor** [comment]
I imaging something like `@=lhist(-10, -10, 10, 0+1);` would still break this, with dodgy memory reads - think we need to also have the semantic analyser ensure we only have literals passed to lhist.

**Comment by ajor** [comment]
Also can you new tests making sure these failure cases (negative numbers, non-literals) are properly rejected please?

**Comment by danobi** [comment]
Yes, let me add some tests on monday.

I also need to read some more code and scratch my head as to how `ast::Unop`s were being tagged as `Type::integer`.

**Comment by ajor** [comment]
Just to be clear: it's intended that unops (and other nodes) get their types set like this. The type represents a node's output, which for negating an integer (in this case) would be another integer. A node being `Type::integer` doesn't mean that it has to be an `ast::Integer`.

**Comment by danobi** [comment]
I see. That makes a lot more sense than having home rolled RTTI.

I've updated the patch with the literal checks. I'm not too happy that it can't support negative values, though. I have two ideas for fixes but I don't like those either:
* Create a pass over the AST that'll fold negative `Unop`s into constants (a lot of work)
* Try a bunch of `dynamic_cast`s in the semantic analyser to manually fold the negative `Unop` into a constant we can error check against (kind of a hack)

File: tests/runtime/call [hunk 1]
```diff
@@ -92,6 +92,16 @@ EXPECT @bytes: *\n[\[(].*
 TIMEOUT 5
 AFTER cat /dev/null
 
+NAME lhist_negative_arg
+RUN bpftrace -v -e 'kretprobe:vfs_read { @bytes = lhist(retval, -10, 10, 1); exit()}'
+EXPECT lhist\(\) expects a integer literal \(integer provided\)
+TIMEOUT 5
+
+NAME lhist_nonliteral_arg
+RUN bpftrace -v -e 'kretprobe:vfs_read { $s = "asdf"; @bytes = lhist(retval, 0, $s, 1); exit()}'
```
Comment (ajor) [inline tests/runtime/call]: We typically do this kind of thing in unit tests in `tests/semantic_analyser.cpp`. No need to run the heavyweight tests for checking invalid programs

Comment (ajor) [reply tests/runtime/call]: We've already got one for `hist()`, but `lhist()` is missing: https://github.com/iovisor/bpftrace/blob/a4bf870ea9733cf7ef048cca367dc3042b7a0cef/tests/semantic_analyser.cpp#L155

**Comment by ajor** [comment]
Negative integer flattening could probably also be done in the parser, with a new rule for `MINUS INT`, but the lexer changes seem like the cleanest way to go about it

**Comment by danobi** [comment]
Ah I misunderstood. I'll add the lexer changes back.

**Comment by danobi** [comment]
@ajor I needed to massage some of the existing tests to make them pass again. They do demonstrate a small problem, though. B/c we now lex integer literals, we lose the ability to assign context dependent associations (if that's the right word for it).

ie `(mytype)arg0+123;` is parsed as `<CAST><INT>` and `(mytype)arg0+ 123;` is parsed as `<CAST><BINOP><INT>`

Do you think the `MINUS INT` rule would be better here?

**Comment by danobi** [comment]
From reading the bison docs, it seems this is a well understood problem. I think the new parser rule is probably the way to do it.

https://www.gnu.org/software/bison/manual/bison.html#Why-Precedence

**Comment by ajor** [comment]
Ah yeah, something like `arg0+123` should be parsed as a binop, so I guess the parser rule is what's needed then

**Comment by danobi** [comment]
@ajor Ok I think this patch finally works as advertised. Would appreciate another review.

File: tests/semantic_analyser.cpp [hunk 2]
```diff
@@ -181,6 +182,12 @@ TEST(semantic_analyser, call_hist)
   test("kprobe:f { hist(); }", 1);
 }
 
+TEST(semantic_analyser, call_lhist)
+{
+  test("kprobe:f { @ = lhist(5, 0, 10, 1); }", 0);
+  test("kprobe:f { @ = lhist(-10, -10, 10, 1); }", 10); // must be positive
```
Comment (ajor) [inline tests/semantic_analyser.cpp]: It would also make sense to check that lhist only works when assigned to a map:
`test("kprobe:f { lhist(5, 0, 10, 1); }", 1);`

File: src/parser.yy [hunk 3]
```diff
@@ -200,7 +201,10 @@ stmt : expr         { $$ = new ast::ExprStatement($1); }
      | var MINUSMINUS  { $$ = new ast::AssignVarStatement($1, new ast::Binop($1, token::MINUS, new ast::Integer(1))); }
      ;
 
-expr : INT             { $$ = new ast::Integer($1); }
+int : MINUS INT    { $$ = new ast::Integer(-1 * $2); }
+    | INT          { $$ = new ast::Integer($1); }
+
```
Comment (ajor) [inline src/parser.yy]:
```suggestion
    ;
```
Guess the final semi colon isn't needed then if this is working... still, I think it should be included for consistency with the rest of the rules in this file

File: tests/semantic_analyser.cpp (hunk 2 above)
Comment (danobi) [reply tests/semantic_analyser.cpp]: Looks like this causes a segfault. Filed #760 for it. Let's fix it in another PR.
